core/classifiers/classifier_utils.py

```diff
@@ -60,15 +61,78 @@ def extract_svm_parameters(clf):
+def unicode_validator_for_classifier_data(classifier_data):
+    """Validates that incoming object contains unicode literal strings.
+
+    Args:
+        classifier_data: *. The trained classifier model data.
+
+    Raises:
+        Exception. If any of the strings in classifier data is not a unicode
+            string.
+    """
+    if isinstance(classifier_data, dict):
+        for k in classifier_data.keys():
             if isinstance(k, str):
                 raise Exception('Expected %s to be unicode but found str.' % k)
-            unicode_validator_for_classifier_data(var[k])
-    if isinstance(var, (list, set, tuple)):
-        for item in var:
+            unicode_validator_for_classifier_data(classifier_data[k])
+    if isinstance(classifier_data, list):
+        for item in classifier_data:
             unicode_validator_for_classifier_data(item)
-    if isinstance(var, str):
-        raise Exception('Expected \'%s\' to be unicode but found str.' % var)
+    if isinstance(classifier_data, str):
+        raise Exception(
+            'Expected \'%s\' to be unicode but found str.' % classifier_data)
+
+
+def convert_float_numbers_to_string_in_classifier_data(classifier_data):
+    """Converts all floating point numbers in classifier data to string.
+
+    The following function iterates through entire classifier data and converts
+    all float values to corresponding string values. At the same time, it also
+    verifies that none of the existing string values are convertible to float
+    values with the help of regex.
+
+    Args:
+        classifier_data: dict|list|string|int|float. The original classifier
+            data which needs conversion of floats to strings.
+
+    Raises:
+        Exception. If any of the string values are convertible to float then
+            an exception is raised to report the error. The classifier data
+            must not include any string values which can be casted to float.
+        Exception. If classifier data contains an object whose type is other
+            than integer, string, dict or list.
+
+    Returns:
+        dict|list|string|int|float. Modified classifier data in which float
+            values are converted into strings.
+    """
+    if isinstance(classifier_data, dict):
+        for k in classifier_data:
+            classifier_data[k] = (
+                convert_float_numbers_to_string_in_classifier_data(
+                    classifier_data[k]))
+        return classifier_data
+    elif isinstance(classifier_data, list):
+        new_list = []
+        for item in classifier_data:
+            new_list.append(
+                convert_float_numbers_to_string_in_classifier_data(item))
+        return new_list
+    elif isinstance(classifier_data, float):
+        return str(classifier_data)
+    elif isinstance(classifier_data, basestring):
+        if re.match(r'^([-+]?\d+\.\d+)$', classifier_data):
+            # A float value must not be stored as a string in
```

Comment:
Is it OK for ints to be stored as strings too? (In a way, ints are floats...)

I'm guessing the answer is no and we want to make the regex a bit stricter.

Also, note that the leading \d might be optional -- e.g. I think ".89" is a valid float. As is 3e10. (Try entering these in Python.)

That said -- it's OK for this "definition of float" to be custom to us. Just recognize that this meanas that the regex is a config param and has to be put in vmconf/feconf etc. and a note saying "keep these in parity between the two systems" for each is required.

Btw, one operational question. What do we do when we find that we need to change this regex -- i.e. what sequence of steps do we take? Is there a way to pause everything (or at least pause sending of stuff to the VM and let the VM clean itself out, and then we do the migration and restart)? We ought to have firm plans and a playbook for that (and probably even do a dry run of that procedure).

Reply:
Updated the regex to `'^([-+]?\\d*\\.\\d+)$|^([-+]?\\d+\\.?\\d*e[-+]?\\d*)$'`. Now it can identify, `[0.123, +0.123, -0.123, .123, +.123, -.123, 3e10, 3e-10, 3e+10, +3e10, -3e10, +3e+10, +3e-10, -3e-10, -3e+10, 0.3e10, -0.3e10, +0.3e10, -0.3e-10, -0.3e+10, +0.3e+10, +0.3e-10]`. I think that contains all variations?

> Btw, one operational question. What do we do when we find that we need to change this regex -- i.e. what sequence of steps do we take? Is there a way to pause everything (or at least pause sending of stuff to the VM and let the VM clean itself out, and then we do the migration and restart)? We ought to have firm plans and a playbook for that (and probably even do a dry run of that procedure).

When we change regex, we stop the VM (no harm in doing that at any point). Update regex here and on Oppia. Start new VM. Before starting new VM we can migrate existing trained classifier model, it required.

Reply:
That list looks pretty good! Does it include 0 and integers? Also try .3e-10 and 00003.1. (I suggested adding a quick test for this, just for reference, so that we can tell at a glance what's included and not included.)

Also, instructions sgtm. Please add them to the comment in vmconf.py so that it's easy to find them when they're needed.

Thanks!